core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java

```diff
@@ -7,149 +7,130 @@
-            // and the image data itself.
-            socketOutputStream.write(MAGIC_NUMBER);
-            socketOutputStream.writeInt(bufferSize);
-            socketOutputStream.write(buffer, 0, bufferSize);
-
-            // Limit the FPS to whatever the dashboard requested
-            int remainingTime = (int) (frameDuration - (System.nanoTime() - startTime));
-            if (remainingTime > 0) {
-              Thread.sleep(remainingTime / 1000000, remainingTime % 1000000);
-            }
-
-            startTime = System.nanoTime();
-          }
-        }
-      } catch (IOException e) {
-        logger.log(Level.WARNING, e.getMessage(), e);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt(); // This is really unnecessary since the thread is
-        // about to exit
-        logger.info("Shutting down camera server");
-        return;
-      } finally {
-        connected = false;
-      }
-    }
-  };
+  private final MjpegServer server;
+  private final CvSource serverSource;
+
+  // Write to the /CameraPublisher table so the MJPEG streams are discoverable by other
+  // applications connected to the same NetworkTable server (eg Shuffleboard)
+  private final ITable cameraPublisherTable = NetworkTable.getTable("/CameraPublisher"); // NOPMD
+  private final ITable ourTable;
+  private final Mat publishMat = new Mat();
+  private long lastFrame = -1;
 
   @SuppressWarnings("JavadocMethod")
   @SuppressFBWarnings(value = "ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD",
       justification = "Do not need to synchronize inside of a constructor")
   public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
-    if (numSteps != 0) {
-      throw new IllegalStateException("Only one instance of PublishVideoOperation may exist");
+    if (numSteps >= MAX_STEP_COUNT) {
+      throw new IllegalStateException(
+          "Only " + MAX_STEP_COUNT + " instances of PublishVideoOperation may exist");
     }
     this.inputSocket = inputSocketFactory.create(SocketHints.Inputs.createMatSocketHint("Image",
         false));
     this.qualitySocket = inputSocketFactory.create(SocketHints.Inputs
         .createNumberSliderSocketHint("Quality", 80, 0, 100));
-    numSteps++;
 
-    serverThread = new Thread(runServer, "Camera Server");
-    serverThread.setDaemon(true);
-    serverThread.start();
+    if (cscoreLoaded) {
+      int ourPort = availablePorts.removeFirst();
+
+      server = new MjpegServer("GRIP video publishing server " + totalStepCount, ourPort);
+      serverSource = new CvSource("GRIP CvSource " + totalStepCount,
+          VideoMode.PixelFormat.kMJPEG, 0, 0, 0);
+      server.setSource(serverSource);
+
+      ourTable = cameraPublisherTable.getSubTable("GRIP-" + totalStepCount);
+      try {
+        InetAddress localHost = InetAddress.getLocalHost();
+        ourTable.putStringArray("streams",
+            new String[]{
+                generateStreamUrl(localHost.getHostName(), ourPort),
```

Comment:
@PeterJohnson Would it be prudent to add a ".local" to the end of the local hostname e.g. `team-190-driverstation.local`? I'm not sure how mDNS would handle it on an FRC field

Reply:
Note `InetAddress.getLocalHost()` may not give you the address you want if the machine has multiple network adapters (https://stackoverflow.com/questions/5813194/inetaddress-getlocalhost-does-not-return-expected-ip-address-from-c-windows-s).  You probably want to use `NetworkInterface.getNetworkInterfaces()` instead.  cscore provides the similar `CameraServerJNI.getNetworkInterfaces()` as well as the NetworkInterfacesChanged event for listening to network interface changes, but at the moment these only work on Linux.  In either case, appending ".local" is almost certainly the wrong thing to do; you should publish the raw IP instead; appending ".local" will require the computer running GRIP to publish its mDNS address, etc.

The streams entry is an array for exactly this reason, so you can e.g. generate an entry for each network address the machine has (e.g. on the robot side we publish with both the USB and Ethernet IP addresses).

Reply:
getNetworkInterfaces works on windows and mac now, so we will be able to use the cscore functions for that. My rewrite will do this